datastore/keyDecodeCodeDup.go

```diff
@@ -0,0 +1,431 @@
+package datastore
+
+import (
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+)
+
+// This code is duplicated from https://github.com/googleapis/google-cloud-go/blob/master/datastore/key.go with the method renamed.
+// decodeToNewKey decodes a key from the opaque representation returned by Encode.
+func decodeToNewKey(encoded string) (*NewFormatKey, error) {
+	// Re-add padding.
+	if m := len(encoded) % 4; m != 0 {
+		encoded += strings.Repeat("=", 4-m)
+	}
+
+	b, err := base64.URLEncoding.DecodeString(encoded)
+	if err != nil {
+		return nil, err
+	}
+
+	pKey := new(PBKey)
+	if err := proto.Unmarshal(b, pKey); err != nil {
+		return nil, err
+	}
+	return cgdProtoToKey(pKey)
+}
+
+// protoToKey decodes a protocol buffer representation of a key into an
+// equivalent *Key object. If the key is invalid, protoToKey will return the
+// invalid key along with ErrInvalidKey.
+func cgdProtoToKey(p *PBKey) (*NewFormatKey, error) {
```

Comment:
what is "cgd"?

maybe this "new" package should be in separate files under an internal package?
would be useful to have some instructions on how the files were constructed, too, given that it's mostly copied code.

Reply:
new package added per your comments